File: 1008. Construct Binary Search Tree from Preorder Traversal/stack.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        stack = [];
        res = None;
        for num in preorder:
            node = TreeNode(num);
            
            if len(stack) == 0:
                stack.append(node);
                res = node;
            elif len(stack) > 0 and num < stack[-1].val:
                stack[-1].left = node;
                stack.append(node)
            else:
                last = None
                while len(stack) > 0 and num > stack[-1].val:
                    last = stack.pop();
                last.right = node;
                stack.append(node)
        return res;

File: 1008. Construct Binary Search Tree from Preorder Traversal/stack.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @return {TreeNode}
 */
 var bstFromPreorder = function(preorder) {
    const stack = [];
    let res;
    
    const top = () => stack[stack.length - 1]
    
    for(const num of preorder){
        const node = new TreeNode(num);
    
        if(stack.length === 0){
            stack.push(node);
            res = node;
        }else if(stack.length && num < top().val){
            top().left = node;
            stack.push(node);
        }else{
            let last;
            while(stack.length && num > top().val){
                last = stack.pop();
            }
            last.right = node
            stack.push(node);
        }
    }
    
    return res;
};